### PR TITLE
dev to live

### PR DESCRIPTION
**Pull request by notefox**
everything was added together
tests were written and ran

### CODE REVIEW

File: src/test/java/de/butbetter/keksys/CalculationValidatorTest.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    private CalculationValidator calc;
+
+    @Test
+    void calculateNegativeNumber() {
+        assertThrows(IllegalArgumentException.class, () -> calc.calculateVAT(-50, 19));
+        assertThrows(IllegalArgumentException.class, () -> calc.calculateVAT(100, -256));
+    }
+
+    @Test
+    void calculateMax() {
```
Comment (Ktechen) [inline src/test/java/de/butbetter/keksys/CalculationValidatorTest.java:22]: Error: 2021-11-07 19:07:10.076  INFO 9692 --- [           main] de.butbetter.keksys.VATCalculator        : 3.4028235E38 + 1.0 = Infinity

File: src/main/java/de/butbetter/keksys/CalculationValidator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public Float calculateVAT(float price, float percent) throws IllegalArgumentException, ArithmeticException {
+        belowThanZero(price, percent);
+        Float priceWithVAT = calculator.calculateVAT(price, percent);
+        priceWithVATInfinityCheck(priceWithVAT);
+        return priceWithVAT;
+    }
+
+    /**
+     * takes price and percent input for VAT Calculation and checks if they or one of them are/is below 0
```
Comment (MathisHTW) [inline src/main/java/de/butbetter/keksys/CalculationValidator.java:25]: checks if one or both of them*

File: src/main/java/de/butbetter/keksys/CalculationValidator.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+     *
+     * @param price   potentially below 0
+     * @param percent potentially below 0
+     * @throws IllegalArgumentException if both or one of the inputs is below 0
```
Comment (MathisHTW) [inline src/main/java/de/butbetter/keksys/CalculationValidator.java:29]: one or both*

File: src/main/java/de/butbetter/keksys/VATCalculator.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+class VATCalculator implements Calculator {
+
+	private static final Logger log = LoggerFactory.getLogger(VATCalculator.class);
+
+	@Override
+	public Float calculateVAT(float price, float percent) {
+		float return_value = ((price / 100) * percent) + price;
```
Comment (MathisHTW) [inline src/main/java/de/butbetter/keksys/VATCalculator.java:14]: rename return_value to priceWithVAT

File: src/main/java/de/butbetter/keksys/Calculator.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package de.butbetter.keksys;
+
+public interface Calculator {
+	/**
+	 * takes a price and a percentage, and returns the value with the given percentage of that value on top
+	 *
```
Comment (MathisHTW) [inline src/main/java/de/butbetter/keksys/Calculator.java:6]: rounded to two decimals

File: src/main/java/de/butbetter/keksys/VATCalculator.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+		log.info(price + " + " + percent + " = " + return_value);
+		return roundTo2DecimalPoints(return_value);
+	}
+
+	/**
+	 * rounds the value up to a given decimal point
+	 */
+	private float roundTo2DecimalPoints(float value) {
```
Comment (MathisHTW) [inline src/main/java/de/butbetter/keksys/VATCalculator.java:22]: could add int decimalPoints = 2, instead of using just 2.
But no biggie i dont care

File: src/test/java/de/butbetter/keksys/CalculationControllerTests.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    private CalculationController calcuationController;
+
+    @Test
+    void contextLoads() {
+        assertThat(calcuationController).isNotNull();
+    }
+
+    @Test
+    void sendParameter14and100() {
+        assertEquals(this.calcuationController.calculation(100, 14).getStatusCode(), HttpStatus.OK);
```
Comment (MathisHTW) [inline src/test/java/de/butbetter/keksys/CalculationControllerTests.java:24]: maybe some more cases like 0 0 / -1 0 / 0 -1 / -1 -1 / Float.MaxValue...

File: src/test/java/de/butbetter/keksys/CalculationValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
+package de.butbetter.keksys;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SpringBootTest
+class CalculationValidatorTest {
```
Comment (MathisHTW) [inline src/test/java/de/butbetter/keksys/CalculationValidatorTest.java:10]: Needs more tests

File: src/test/java/de/butbetter/keksys/VATCalculatorTest.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class VATCalculatorTest {
+
+	private VATCalculator calc;
+
+	@BeforeEach
+	void setup() {
+		calc = new VATCalculator();
+	}
```
Comment (MathisHTW) [inline src/test/java/de/butbetter/keksys/VATCalculatorTest.java:15]: test some edge cases like Float.MinValue .../ 0 0 / Float.MaxValue

**Review by MathisHTW — CHANGES_REQUESTED** [review]

File: src/main/java/de/butbetter/keksys/VATCalculator.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+@Service
+class VATCalculator implements Calculator {
+
+	private static final Logger log = LoggerFactory.getLogger(VATCalculator.class);
+
+	@Override
+	public Float calculateVAT(float price, float percent) {
+		float return_value = ((price / 100) * percent) + price;
+		log.info(price + " + " + percent + " = " + return_value);
+		return roundTo2DecimalPoints(return_value);
```
Comment (MathisHTW) [inline src/main/java/de/butbetter/keksys/VATCalculator.java:16]: Rounding to two decimals is fine for smaller values
But for really big values this is **fatal**:
Float Max = 3.4028235E38
But with this solution:
calculateVAT(Float.MaxValue,1) which should also be = 3.4028235E38, is just 2.1474836E7!

WE NEED TO FIND A DIFFERENT SOLUTION FOR THIS

**Review by MathisHTW — CHANGES_REQUESTED** [review]

File: src/main/java/de/butbetter/keksys/CalculationController.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+@RestController
+@RequestMapping("calc/v1/")
```
[the remark this answers is not in the document]
Comment (Ktechen) [reply src/main/java/de/butbetter/keksys/CalculationController.java:18]: Are you have any error with the requestmapping?

File: src/main/java/de/butbetter/keksys/CalculationController.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+     * @return ResponseEntity if all correct you will get HttpStatus.OK and VAT result as json
+     */
+    @GetMapping("/VAT")
+    public ResponseEntity<Object> calculation(
+            @RequestParam(value = "price") float price,
+            @RequestParam(value = "percent") float percent
+    ) {
+
+        try {
+            Float result = calculationValidator.calculateVAT(price, percent);
```
[the remark this answers is not in the document]
Comment (Ktechen) [reply src/main/java/de/butbetter/keksys/CalculationController.java:41]: Maybe resultOfVAT?

File: src/main/java/de/butbetter/keksys/CalculationController.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+            logger.info(String.format("(%.2f / 100) * %.2f) + %.2f = %.2f", price, percent, price, result));
+            Map<String, Float> jsonObject = new HashMap<>();
+            jsonObject.put(JSON_VAT_NAME, result);
+
+            return new ResponseEntity<>(jsonObject, HttpStatus.OK);
+
+        } catch (IllegalArgumentException | ArithmeticException e) {
```
[the remark this answers is not in the document]
Comment (Ktechen) [reply src/main/java/de/butbetter/keksys/CalculationController.java:48]: No Errormessage to Customer I think so

File: src/test/java/de/butbetter/keksys/CalculationValidatorTest.java (hunk 8 above)
Comment (Ktechen) [reply src/test/java/de/butbetter/keksys/CalculationValidatorTest.java:11]: I don't think so

File: src/test/java/de/butbetter/keksys/VATCalculatorTest.java (hunk 9 above)
Comment (Ktechen) [reply src/test/java/de/butbetter/keksys/VATCalculatorTest.java:15]: I don't know what are you mean?

File: src/test/java/de/butbetter/keksys/CalculationValidatorTest.java (hunk 8 above)
Comment (MathisHTW) [reply src/test/java/de/butbetter/keksys/CalculationValidatorTest.java:11]: 🥇

File: src/main/java/de/butbetter/keksys/CalculationController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/CalculationController.java:18]: no but i thought it would crash, if it doesnt nvm

File: src/main/java/de/butbetter/keksys/CalculationController.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/CalculationController.java:41]: Maybe resultPriceOFWithVAT?

File: src/main/java/de/butbetter/keksys/CalculationController.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/CalculationController.java:48]: ok, i thought it would be a good idea, but then nvm

File: src/main/java/de/butbetter/keksys/VATCalculator.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+	}
+
+	/**
+	 * rounds the value up to a given decimal point
+	 */
+	private float roundTo2DecimalPoints(float value) {
+		return new BigDecimal(value).setScale(2, RoundingMode.HALF_UP).floatValue();
```
Comment (MathisHTW) [inline src/main/java/de/butbetter/keksys/VATCalculator.java:27]: You cannot store infinity in a BigDecimal, 
this throws a NumberFormatException

**Review by MathisHTW — CHANGES_REQUESTED** [review]

File: src/test/java/de/butbetter/keksys/VATCalculatorTest.java (hunk 9 above)
Comment (MathisHTW) [reply src/test/java/de/butbetter/keksys/VATCalculatorTest.java:15]: just some extra test cases with maybe some parameters like:
Float.MinValue .../ 0 0 / Float.MaxValue

Float.MaxValue and 1 MAYBE? lel

File: src/main/java/de/butbetter/keksys/CalculationController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (Ktechen) [reply src/main/java/de/butbetter/keksys/CalculationController.java:18]: Explain what do you mean nvm(Non-Volatile Memory) ?

File: src/main/java/de/butbetter/keksys/CalculationController.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ktechen) [reply src/main/java/de/butbetter/keksys/CalculationController.java:41]: Maybe not

File: src/main/java/de/butbetter/keksys/VATCalculator.java (hunk 14 above)
Comment (Ktechen) [reply src/main/java/de/butbetter/keksys/VATCalculator.java:27]: Issue detected: https://app.codacy.com/gh/But-Better/CalcApi/pullRequest?prid=8370557

File: src/test/java/de/butbetter/keksys/CalculationControllerTests.java (hunk 7 above)
Comment (Ktechen) [reply src/test/java/de/butbetter/keksys/CalculationControllerTests.java:24]: okay

File: src/test/java/de/butbetter/keksys/CalculationValidatorTest.java (hunk 8 above)
Comment (Ktechen) [reply src/test/java/de/butbetter/keksys/CalculationValidatorTest.java:11]: 🥇

File: src/test/java/de/butbetter/keksys/VATCalculatorTest.java (hunk 9 above)
Comment (Ktechen) [reply src/test/java/de/butbetter/keksys/VATCalculatorTest.java:15]: Next one try to explain a explain with assert explain not like this .../ 0 0 /  👍  💯 
More like this:
```java
 assertEquals(0,calc.calculateVAT(0,0));
```

File: src/main/java/de/butbetter/keksys/CalculationController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/CalculationController.java:18]: nvm = never mind ^^

File: src/main/java/de/butbetter/keksys/VATCalculator.java (hunk 14 above)
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/VATCalculator.java:29]: VATcalc now returns infinity like before

File: src/test/java/de/butbetter/keksys/CalculationValidatorTest.java (hunk 8 above)
Comment (MathisHTW) [reply src/test/java/de/butbetter/keksys/CalculationValidatorTest.java:11]: more test cases were added

File: src/test/java/de/butbetter/keksys/CalculationControllerTests.java (hunk 7 above)
Comment (MathisHTW) [reply src/test/java/de/butbetter/keksys/CalculationControllerTests.java:24]: more test cases were added

File: src/main/java/de/butbetter/keksys/Calculator.java (hunk 5 above)
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/Calculator.java:6]: comment added

File: src/main/java/de/butbetter/keksys/CalculationValidator.java (hunk 3 above)
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/CalculationValidator.java:29]: updated doku

File: src/main/java/de/butbetter/keksys/CalculationValidator.java (hunk 2 above)
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/CalculationValidator.java:25]: updated doku

File: src/test/java/de/butbetter/keksys/CalculationValidatorTest.java (hunk 1 above)
Comment (MathisHTW) [reply src/test/java/de/butbetter/keksys/CalculationValidatorTest.java]: Error was caused by unexpected Exception and is resolved in e42deb3

File: src/main/java/de/butbetter/keksys/VATCalculator.java (hunk 4 above)
Comment (MathisHTW) [reply src/main/java/de/butbetter/keksys/VATCalculator.java:18]: name was updated

File: src/main/java/de/butbetter/keksys/VATCalculator.java (hunk 6 above)
Comment (notefox) [reply src/main/java/de/butbetter/keksys/VATCalculator.java:27]: could do later

File: src/main/java/de/butbetter/keksys/VATCalculator.java (hunk 14 above)
Comment (notefox) [reply src/main/java/de/butbetter/keksys/VATCalculator.java:29]: wouldn't use this kind of check, since this is partially a code copy of the validator,
I'd rather try and catch the exception thrown and turn it into a ArithmeticException instead, which should be handled by either the Validator or the Controller at some point already

Comment (notefox) [reply src/main/java/de/butbetter/keksys/VATCalculator.java:29]: or we catch the FormatException, and return the value anyways
all 3 could work

File: src/main/java/de/butbetter/keksys/CalculationController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (notefox) [reply src/main/java/de/butbetter/keksys/CalculationController.java:18]: I've not seen any error coming for there YET, seems like Spring Boot just doesn't care
I'd just ignore it
 
we won't deal with that side of the application anyways, since all the building and handling requests from a browser directly is being done on the Spring Boot Side, and if it can handle it
it's fine

File: src/main/java/de/butbetter/keksys/CalculationController.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (notefox) [reply src/main/java/de/butbetter/keksys/CalculationController.java:48]: That idea with returning an error messages in form of a json is not far of from what the business standart actually is
just look at an example from a [Google API Error Documentation](https://developers.google.com/webmaster-tools/search-console-api-original/v3/errors)

File: src/main/java/de/butbetter/keksys/VATCalculator.java (hunk 14 above)
Comment (notefox) [reply src/main/java/de/butbetter/keksys/VATCalculator.java]: decided to catch exception and just give back value when no rounding can be applied

**Review by MathisHTW — APPROVED** [review]
approve

**Review by notefox — COMMENTED** [review]
works for me

**Review by Ktechen — APPROVED** [review]
